### PR TITLE
Removal of Unnecessary Overloads + Extension method tweaks.

### PR DESCRIPTION
**Pull request by AdamSpeight2008**

### CODE REVIEW

**Comment by mikedn** [comment]
I don't know if matters here but removing the non-params overloads leads to an unnecessary empty array allocation if no additional arguments are provided. The DiagnosticInfo class that ends up storing the args is prepared to deal with null arrays and that may be a sign that the overloads were intentional.

As for IsBetween, well, that sounds like a reasonable attempt at killing the lexer performance :smile:

**Comment by AdamSpeight2008** [comment]
@mikedn Why would it create an array? If no parameters are passed should the value of the params be `null` since an array is class type.

Of all the cases *(VS can find)* it is passed only one argument a string, an array see like to much.

**Comment by mikedn** [comment]
It creates an array of 0 length because that's what the C# specification requires. 

I searched the code a bit and there are a couple of uses of the non-params overloads. Even if there are no concerns about allocation I'd probably keep those overloads because they avoid bloating the call sites with useless code for the array allocation.

**Comment by AdamSpeight2008** [comment]
Why not optional argument?

     protected void AddError(int position, int width, ErrorCode code, object[] args = null)
Plus an overload for the case of `string`

     protected void AddError(int position, int width, ErrorCode code, string arg)

See commit 65d3d0ab57805b70a3f4cedfce6cd8079f839174

**Comment by mikedn** [comment]
Seems like too much trouble. Avoiding allocating an empty array at the expense of a couple of null checks and trivial overloads is one thing. Avoiding allocating an array with a single element at the expense of additional overloads and more code to deal with this special case is another thing and should only be done if there's evidence that it improves performance.

Also, your PR contains a bunch of unrelated and unnecessary formatting changes. In particular, you seem to indent the code with 2 spaces while this project is using 4.

**Comment by AdamSpeight2008** [comment]
Then if that's the case why the class [OneOrMany](https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/InternalUtilities/OneOrMany.cs)?

**Comment by mikedn** [comment]
Presumably because someone has done the necessary analysis and decided that it's worth the trouble. I've never said that one should never do such an optimization, I said that such optimizations should only be done when they're really needed.

You started your PR with the goal to cleanup some code and when I pointed out that a particular change might be a de-optimization you changed the goal from cleanup to performance improvement, even if there's no evidence that the improvement is necessary.

**Comment by paulomorgado** [comment]
Changing:

           sb.AppendFormat("Summary: {0}\n", Kind);

into:

            sb.AppendLine($"Summary: {Kind}");

might seem pretty and modern but, in reality, it's changing into:

            sb.AppendLine(string.Formt("Summary: {0}", Kind));

Behind the curtain, `string.Format` uses a `StriungBuilder`. That would mean an allocation of a new `StringBuilder` when one is already being used.

**Comment by paulomorgado** [comment]
Is there anywhere else in the Roslyn code where optional arguments are being used on public or protected methods?

File: src/Compilers/CSharp/Portable/Binder/Semantics/Operators/OperatorKindExtensions.cs [hunk 1]
```diff
@@ -152,7 +152,7 @@ public static BinaryOperatorKind WithType(this BinaryOperatorKind kind, SpecialT
                 case SpecialType.System_UInt64:
                     return kind | BinaryOperatorKind.ULong;
                 default:
-                    Debug.Assert(false, "Unexpected binary operator type.");
+                    Debug.Assert(false, _Unexpected);
```
Comment (thorgeirk11) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Operators/OperatorKindExtensions.cs]: The previous error stated that this was a binary operator. This is now incorrect.

**Comment by AdamSpeight2008** [comment]
I'm still trying to get used to Git. I thought that a pull request would only pull up to a specific point of time, not subsequent commits.

File: src/Compilers/CSharp/Portable/Errors/SyntaxDiagnosticInfo.cs [hunk 2]
```diff
@@ -11,7 +11,15 @@ internal class SyntaxDiagnosticInfo : DiagnosticInfo
...
 
-        internal SyntaxDiagnosticInfo(int offset, int width, ErrorCode code, params object[] args)
+
+    internal SyntaxDiagnosticInfo(int offset, int width, ErrorCode code, string arg)
+    : base(CSharp.MessageProvider.Instance, (int)code, arg)
+    {
+      Debug.Assert(width >= 0);
+      this.Offset = offset;
+      this.Width = width;
+    }
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Errors/SyntaxDiagnosticInfo.cs]: Nit: blank line separator

File: src/Compilers/CSharp/Portable/Errors/XmlSyntaxDiagnosticInfo.cs [hunk 3]
```diff
@@ -9,8 +9,11 @@ namespace Microsoft.CodeAnalysis.CSharp
     internal sealed class XmlSyntaxDiagnosticInfo : SyntaxDiagnosticInfo
     {
         private readonly XmlParseErrorCode xmlErrorCode;
-
-        internal XmlSyntaxDiagnosticInfo(XmlParseErrorCode code, params object[] args)
+    internal XmlSyntaxDiagnosticInfo(XmlParseErrorCode code, string arg)
+         : this(0, 0, code, arg)
+    {
+    }
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Errors/XmlSyntaxDiagnosticInfo.cs]: Nit: Need a blank line separator here

File: src/Compilers/CSharp/Portable/Parser/AbstractLexer.cs [hunk 4]
```diff
@@ -60,38 +60,28 @@ protected SyntaxDiagnosticInfo[] GetErrors(int leadingTriviaWidth)
...
-        }
-
-        protected void AddError(int position, int width, ErrorCode code, params object[] args)
+    protected void AddError(int position, int width, ErrorCode code, object[] args = null)
         {
             this.AddError(this.MakeError(position, width, code, args));
         }
 
-        protected void AddError(int position, int width, XmlParseErrorCode code, params object[] args)
+        protected void AddError(int position, int width, XmlParseErrorCode code, object[] args = null)
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Parser/AbstractLexer.cs]: What problem does this change solve?

File: src/Compilers/CSharp/Portable/Parser/AbstractLexer.cs [hunk 5]
```diff
@@ -109,19 +99,21 @@ protected void AddError(SyntaxDiagnosticInfo error)
...
+      return new SyntaxDiagnosticInfo(offset, width, code, arg);
+    }
+
+
+    protected SyntaxDiagnosticInfo MakeError(int position, int width, ErrorCode code, object[] args = null)
         {
             int offset = GetLexemeOffsetFromPosition(position);
-            return new SyntaxDiagnosticInfo(offset, width, code, args);
+            return new SyntaxDiagnosticInfo( offset, width, code, args);
         }
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Parser/AbstractLexer.cs]: Nit: extra space

File: src/Compilers/CSharp/Portable/Utilities/GenericExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,24 @@
+﻿using System;
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Utilities/GenericExtensions.cs]: Make sure to include the copyright notice at the top of new files

File: src/Compilers/CSharp/Portable/Utilities/GenericExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis
+{
+  public enum Clusivity : int
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Utilities/GenericExtensions.cs]: I think we should use a different name here.  I agree that the term clusivity has the appropriate meaning for this use case but I don't think it's a common enough word for a type name.  

Admittedly I had to look up this word :smile:

File: src/Compilers/CSharp/Portable/Utilities/GenericExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  {
+    @In = 0,
+    @Ex = -1
+  }
+  public static class GenericExtensions
+  {
+
+
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Utilities/GenericExtensions.cs]: Nit: double blank line

File: src/Compilers/CSharp/Portable/Utilities/GenericExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis
+{
+  public enum Clusivity : int
+  {
+    @In = 0,
+    @Ex = -1
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Utilities/GenericExtensions.cs]: Don't think the `@` are needed here.

File: src/Compilers/Core/Portable/Diagnostic/CommonDiagnosticComparer.cs [hunk 10]
```diff
@@ -14,26 +14,14 @@ internal sealed class CommonDiagnosticComparer : IEqualityComparer<Diagnostic>
...
-                return true;
-            }
-
-            if (x == null || y == null)
-            {
-                return false;
-            }
-
+            if (object.ReferenceEquals(x, y))  return true;
+            if (x == null || y == null)        return false;
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/Diagnostic/CommonDiagnosticComparer.cs]: There are a couple of issues here:

The first is that single line `if` statements aren't allowed under the [.NET foundation coding style](https://github.com/dotnet/corefx/wiki/Contributing#c-coding-style).  This is the style that Roslyn is following now for all of its code.  

The second is that in general we are going to be reluctant to take style only changes unless it either:

1. Significantly improves the readability of the code in question. 
2. Changes code to conform to the stated guidelines which previously did not.

**Comment by jaredpar** [comment]
Thank you for the contribution!

A significant portion of this change is simply altering the style of the existing code.  In general we are going to be reluctant to take such changes.  They generally represent churn in the code base that don't provide concrete value. 

Overall our code base now follows the [.NET Foundation contributing guidelines](https://github.com/dotnet/corefx/wiki/Contributing#c-coding-style) including their coding style requirements.  Much of the code didn't follow this when originally posted but over the last few weeks it has all been migrated to meet those guidelines.  The code will continue to follow those guidelines going forward.  

Style changes will generally be taken only when it:

- Significantly improves the readability of the code in question.
- Changes code to conform to the stated guidelines which previously did not.

**Comment by AdamSpeight2008** [comment]
@jaredpar The single line if's do improve readability.
 Maybe allow them if it is small ( eg with the total line width of 80 chars?)
Eg Guards 
```c#
if( arg0 == null ) throw new NullArgumentException( nameof(arg0));
if( arg1 == null ) throw new NullArgumentException( nameof(arg1));
```
More of guards and you are talk in the region of half the screen height, before you even get to the "good stuff" of the method. *not all of have huge monitors, some laptops has more horizontal space than vertical*

Simple returns
```c#
if( x == 0 ) return "Zero";
```
SImple Cases
```c#
select ( value )
{
  case 0:  return "Zero";
  case 1:  return "One";
  case 2:  return "Two";
  case 3:  return "Three";
  case 4:  return "Four";
  case 5:  return "Five";
  case 6:  return "Six";
  case 7:  return "Seven";
  case 8:  return "Eight";
  case 9:  return "Nine";
  default:
    return "shrug";
}
```
15 LoC vs 25 (separate lines )  or 35 if in  require the breaks.
The 4 space indentation is too wide, it should be 2. Considering the each block / scope get indented.

Now consider the case where pattern-matching is included, the code will get both wider and longer (in terms of structural layout).

**Comment by jaredpar** [comment]
@AdamSpeight2008 whether or not single line ifs, or most other style decisions, improve readability is subjective.  What is preferable to one developer may not be preferable to the other.  The important aspect of style is to keep it consistent across the code base.  

In this specific case though single line ifs are not permitted by the coding guidelines.

**Comment by sharwell** [comment]
@AdamSpeight2008 You submitted a pull request from your `master` branch. This means any subsequent commits you make to the `master` branch in your fork while this pull request is open will automatically appear here. To make things much easier, you should create a new branch in your fork and use it to perform one specific change (e.g. changing instances of `AppendFormat` to use string interpolation). You can then send a pull request to merge that branch and it can be evaluated independently of any other type of changes you want to propose.

To start working on a different proposed change, such as creating new constants for string literals that appear in code, go back to the latest `master` commit and create *another* new branch for that change.

Otherwise, when you send a bunch of different types of changes in a single pull request, it has a high likelihood of getting rejected if people disagree with one of the changes, *even if* they agree with the majority.

**Comment by AdamSpeight2008** [comment]
Note: The original repo for this PR as been removed.

**Pull request closed without merging**